### PR TITLE
Solved Gaussian value range #2383

### LINKED ISSUE #2383: Why is `gaussian` so opinionated about the image value range
## Description

Trying to apply `skimage.filters.gaussian` to a floating image may result in a `ValueError` if any value is outside [0, 1]. For `gaussian` it shouldn't matter what values are in the image.

```pytb
In [1]: import numpy as np
In [2]: from skimage.filters import gaussian
In [3]: img = np.array([[10.0, -10.0], [-4, 3]], dtype=np.float32)
In [4]: gaussian(img, 1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-15-b8596e79c5dc> in <module>()
----> 1 gaussian(img, 1)

/home/oli/Projects/pbl2/env/lib/python3.5/site-packages/skimage/filters/_gaussian.py in gaussian(image, sigma, output, mode, cval, multichannel)
    102         if len(sigma) != image.ndim:
    103             sigma = np.concatenate((np.asarray(sigma), [0]))
--> 104     image = img_as_float(image)
    105     return ndi.gaussian_filter(image, sigma, mode=mode, cval=cval)

/home/oli/Projects/pbl2/env/lib/python3.5/site-packages/skimage/util/dtype.py in img_as_float(image, force_copy)
    299
    300     """
--> 301     return convert(image, np.float64, force_copy)
    302
    303

/home/oli/Projects/pbl2/env/lib/python3.5/site-packages/skimage/util/dtype.py in convert(image, dtype, force_copy, uniform)
    203     if kind_in == 'f':
    204         if np.min(image) < -1.0 or np.max(image) > 1.0:
--> 205             raise ValueError("Images of type float must be between -1 and 1.")
    206         if kind == 'f':
    207             # floating point -> floating point

ValueError: Images of type float must be between -1 and 1.
```

## Way to reproduce

Simply run the code snippet:

```python
import numpy as np
from skimage.filters import gaussian
img = np.array([[10.0, -10.0], [-4, 3]], dtype=np.float32)
gaussian(img, 1)
```

### PR DESCRIPTION
**Pull request by mesejo**
## Description
Solved the bug described in the ticket. Created a new function based on existing ones in skimage.transform._geometry. Add preserve_range parameter and corresponding documentation. In order to solved the problem the img_as_float call was replaced _convert_input.
Test was added in test_gaussian.py

closes #2383

### CODE REVIEW

**Comment by sciunto** [comment]
@mesejo Many thanks for your contribution!

@jni, from your comments in the bug ticket, would you prefer a default value set to True?

**Comment by sciunto** [comment]
In `skimage/transform/_geometric.py`, there is also a function _convert_warp_input(image, preserve_range).

I think we should put a function in common in skimage/_shared

@mesejo Would you mind doing that and fixing functions calls in skimage.transform?

**Comment by mesejo** [comment]
@sciunto I don't mind, in which file of the _shared folder should I put the common function?

File: skimage/_shared/utils.py [hunk 1]
```diff
@@ -208,3 +209,12 @@ def check_random_state(seed):
         return seed
     raise ValueError('%r cannot be used to seed a numpy.random.RandomState'
                      ' instance' % seed)
+
+
+def _convert_input(image, preserve_range):
```
Comment (sciunto) [inline skimage/_shared/utils.py]: could you remove the underscore please? The folder is already private.

File: skimage/_shared/utils.py [hunk 2]
```diff
@@ -208,3 +209,12 @@ def check_random_state(seed):
...
+    """Convert input image to double image with the appropriate range."""
```
Comment (sciunto) [inline skimage/_shared/utils.py]: Could you write a Parameters section, like above please?

File: skimage/transform/_geometric.py [hunk 3]
```diff
@@ -1197,7 +1197,7 @@ def warp_coords(coord_map, shape, dtype=np.float64):
     return coords
 
 
-def _convert_warp_input(image, preserve_range):
+def _convert_input(image, preserve_range):
```
Comment (sciunto) [inline skimage/transform/_geometric.py]: Thus function must be removed, since it's now in _shared.

File: skimage/_shared/utils.py [hunk 4]
```diff
@@ -208,3 +209,26 @@ def check_random_state(seed):
         return seed
     raise ValueError('%r cannot be used to seed a numpy.random.RandomState'
                      ' instance' % seed)
+
+
+def convert_input(image, preserve_range):
```
Comment (blink1073) [inline skimage/_shared/utils.py]: This is too vague a name for a utility function, perhaps `convert_to_float`?

**Comment by blink1073** [comment]
Hi @mesejo, I had one inline comment, and this unfortunately needs a [rebase](http://scikit-image.org/docs/stable/contribute.html#divergence-between-upstream-master-and-your-feature-branch).

**Comment by sciunto** [comment]
kind ping to @mesejo

**Comment by mesejo** [comment]
Hi, I have a question I am trying to do a rebase from master but it seems its overwriting my changes. Does it really needs a rebase??

**Comment by sciunto** [comment]
Yes, we do. There is a conflict with master.

**Comment by mesejo** [comment]
So does it means that i have to redo the changes. Could yo give me any pointers to avoid redoing that?

**Comment by sciunto** [comment]
Good! I think that @blink1073 can now do the second review. :)

**Comment by blink1073** [comment]
Works for me, thanks @mesejo!

**Pull request merged**